// @adobe/target-node-client/lib/target/batch-mbox-v2-api.js
const https = require("https");
const querystring = require("querystring");
const DEFAULT_AGENT = new https.Agent({
  keepAlive: true
});
const POST = "POST";
const SECURE_PROTOCOL = "https:";
const UTF8_ENCODING = "utf8";
const ERROR = "error";
const PARSE_ERROR = "parse_error";
const ABORT = "abort";
const TIMEOUT = "timeout";
const ABORTED = "Request aborted";
const TIMED_OUT = "Request timed out";

const RESPONSE_EVENT = "response";
const DATA_EVENT = "data";
const END_EVENT = "end";
const ABORT_EVENT = "abort";
const TIMEOUT_EVENT = "timeout";
const ERROR_EVENT = "error";
const SOCKET_EVENT = "socket";
const LOOKUP_EVENT = "lookup";
const CONNECT_EVENT = "connect";
const SECURE_CONNECT_EVENT = "secureConnect";

const NS_PER_SEC = 1e9;
const MS_PER_NS = 1e6;

function getNanoSeconds() {
  const hr = process.hrtime();

  return hr[0] * NS_PER_SEC + hr[1];
}

const UP_TIME = process.uptime() * NS_PER_SEC;
const MODULE_LOAD_TIME = getNanoSeconds();
const NODE_LOAD_TIME = MODULE_LOAD_TIME - UP_TIME;

function now() {
  return (getNanoSeconds() - NODE_LOAD_TIME) / MS_PER_NS;
}

function creatError(status, message, timings) {
  return { status, message, timings };
}

function createSuccess(data, response, timings) {
  return { data, response, timings };
}

function getTimings(timings) {
  if (!timings.socket) {
    timings.socket = 0;
  }

  if (!timings.lookup) {
    timings.lookup = timings.socket;
  }

  if (!timings.connect) {
    timings.connect = timings.lookup;
  }

  if (!timings.secureConnect) {
    timings.secureConnect = timings.connect;
  }

  if (!timings.response) {
    timings.response = timings.secureConnect;
  }

  return {
    wait: timings.socket,
    dns: timings.lookup - timings.socket,
    tcp: timings.connect - timings.lookup,
    tls: timings.secureConnect - timings.connect,
    firstByte: timings.response - timings.secureConnect,
    download: timings.end - timings.response,
    total: timings.end
  };
}

function createRequestOptions(host, path, query, headers) {
  return {
    method: POST,
    protocol: SECURE_PROTOCOL,
    host,
    path: `${path}?${querystring.stringify(query)}`,
    headers,
    agent: DEFAULT_AGENT
  };
}

function executeRequest(options, body, timeout, callback) {
  const timings = {};
  const chunks = [];
  const startTimeNow = now();
  const request = https.request(options, res => {
    res.setEncoding(UTF8_ENCODING);
  });
  const onLookup = () => {
    timings.lookup = now() - startTimeNow;
  };
  const onConnect = () => {
    timings.connect = now() - startTimeNow;
  };
  const onSecureConnect = () => {
    timings.secureConnect = now() - startTimeNow;
  };
  const onError = err => {
    callback(creatError(ERROR, err, getTimings(timings)));

    if (request.socket) {
      request.socket.removeListener(LOOKUP_EVENT, onLookup);
      request.socket.removeListener(CONNECT_EVENT, onConnect);
      request.socket.removeListener(SECURE_CONNECT_EVENT, onSecureConnect);
    }
  };
  const onSocket = socket => {
    timings.socket = now() - startTimeNow;

    if (socket._connecting || socket.connecting) {
      socket.once(LOOKUP_EVENT, onLookup);
      socket.once(CONNECT_EVENT, onConnect);
      socket.once(SECURE_CONNECT_EVENT, onSecureConnect);
      socket.once(ERROR_EVENT, onError);
    }
  };
  const onResponse = response => {
    timings.response = now() - startTimeNow;

    response.on(DATA_EVENT, chunk => {
      chunks.push(chunk);
    });

    response.once(END_EVENT, () => {
      const content = chunks.join("");
      timings.end = now() - startTimeNow;

      if (response.statusCode !== 200) {
        callback(creatError(ERROR, content, getTimings(timings)));
        return;
      }

      try {
        const data = JSON.parse(content);

        callback(null, createSuccess(data, content, getTimings(timings)));
      } catch (error) {
        callback(creatError(PARSE_ERROR, error, getTimings(timings)));
      }
    });
  };

  request.once(RESPONSE_EVENT, onResponse);
  request.once(SOCKET_EVENT, onSocket);
  request.once(ABORT_EVENT, () =>
    callback(creatError(ABORT, ABORTED, getTimings(timings)))
  );
  request.once(TIMEOUT_EVENT, () =>
    callback(creatError(TIMEOUT, TIMED_OUT, getTimings(timings)))
  );
  request.once(ERROR_EVENT, onError);

  request.setTimeout(timeout);
  request.write(Buffer.from(JSON.stringify(body)), UTF8_ENCODING);
  request.end();
}

/**
 * @module ApiClient
 * @version 1.0.0
 */

/**
 * Manages low level client-server communications, parameter marshalling, etc. There should not be any need for an
 * application to use this class directly - the *Api and model classes provide the public API for the service. The
 * contents of this file should be regarded as internal but are documented for completeness.
 * @alias module:ApiClient
 * @class
 */
class ApiClient {
  constructor(opts) {
    opts = opts || {};

    /**
     * The default HTTP timeout for all API calls.
     * @type {Number}
     * @default 10000
     */
    this.timeout = opts.timeout || 10000;
  }

  /**
   * Returns a string representation for an actual parameter.
   * @param param The actual parameter.
   * @returns {String} The string representation of <code>param</code>.
   */
  paramToString(param) {
    if (param == undefined || param == null) {
      return "";
    }
    if (param instanceof Date) {
      return param.toJSON();
    }

    return param.toString();
  }

  /**
   * Normalizes parameter values:
   * <ul>
   * <li>remove nils</li>
   * <li>keep files and arrays</li>
   * <li>format to string with `paramToString` for other cases</li>
   * </ul>
   * @param {Object.<String, Object>} params The parameters as object properties.
   * @returns {Object.<String, Object>} normalized parameters.
   */
  normalizeParams(params) {
    var newParams = {};
    for (var key in params) {
      if (
        params.hasOwnProperty(key) &&
        params[key] != undefined &&
        params[key] != null
      ) {
        var value = params[key];

        newParams[key] = this.paramToString(value);
      }
    }

    return newParams;
  }

  /**
   * Deserializes an HTTP response body into a value of the specified type.
   * @param {Object} response A SuperAgent response object.
   * @param {(String|Array.<String>|Object.<String, Object>|Function)} returnType The type to return. Pass a string for simple types
   * or the constructor function for a complex type. Pass an array containing the type name to return an array of that type. To
   * return an object, pass an object with one property whose name is the key type and whose value is the corresponding value type:
   * all properties on <code>data<code> will be converted to this type.
   * @returns A value of the specified type.
   */
  deserialize(response, returnType) {
    if (response == null || returnType == null) {
      return null;
    }

    return ApiClient.convertToType(response, returnType);
  }

  /**
   * Invokes the REST service using the supplied settings and parameters.
   * @param {host} host The server host.
   * @param {String} path The base URL to invoke.
   * @param {Object.<String, Object>} queryParams A map of query parameters and their values.
   * @param {Object.<String, Object>} headerParams A map of header parameters and their values.
   * @param {Object} bodyParam The value to pass as the request body.
   * @param {(String|Array|ObjectFunction)} returnType The required type to return; can be a string for simple types or the
   * constructor for a complex type.
   * @returns {Promise} A {@link https://www.promisejs.org/|Promise} object.
   */
  callApi(host, path, queryParams, headerParams, bodyParam, returnType) {
    const options = createRequestOptions(host, path, queryParams, headerParams);

    return new Promise((resolve, reject) => {
      executeRequest(options, bodyParam, this.timeout, (err, response) => {
        if (err) {
          reject(err);
          return;
        }

        const result = response;
        result.data = this.deserialize(result.data, returnType);

        resolve(result);
      });
    });
  }

  /**
   * Parses an ISO-8601 string representation of a date value.
   * @param {String} str The date value as a string.
   * @returns {Date} The parsed date object.
   */
  static parseDate(str) {
    return new Date(str);
  }

  /**
   * Converts a value to the specified type.
   * @param {(String|Object)} data The data to convert, as a string or object.
   * @param {(String|Array.<String>|Object.<String, Object>|Function)} type The type to return. Pass a string for simple types
   * or the constructor function for a complex type. Pass an array containing the type name to return an array of that type. To
   * return an object, pass an object with one property whose name is the key type and whose value is the corresponding value type:
   * all properties on <code>data<code> will be converted to this type.
   * @returns An instance of the specified type or null or undefined if data is null or undefined.
   */
  static convertToType(data, type) {
    if (data === null || data === undefined) return data;

    switch (type) {
      case "Boolean":
        return Boolean(data);
      case "Integer":
        return parseInt(data, 10);
      case "Number":
        return parseFloat(data);
      case "String":
        return String(data);
      case "Date":
        return ApiClient.parseDate(String(data));
      default:
        if (type === Object) {
          // generic object, return directly
          return data;
        } else if (typeof type === "function") {
          // for model type like: User
          return type.constructFromObject(data);
        } else if (Array.isArray(type)) {
          // for array type like: ['String']
          var itemType = type[0];

          return data.map(item => {
            return ApiClient.convertToType(item, itemType);
          });
        } else if (typeof type === "object") {
          // for plain object type like: {'String': 'Integer'}
          var keyType, valueType;
          for (var k in type) {
            if (type.hasOwnProperty(k)) {
              keyType = k;
              valueType = type[k];
              break;
            }
          }

          var result = {};
          for (var k in data) {
            if (data.hasOwnProperty(k)) {
              var key = ApiClient.convertToType(k, keyType);
              var value = ApiClient.convertToType(data[k], valueType);
              result[key] = value;
            }
          }

          return result;
        } else {
          // for unknown type, return the data directly
          return data;
        }
    }
  }

  /**
   * Constructs a new map or array model from REST data.
   * @param data {Object|Array} The REST data.
   * @param obj {Object|Array} The target object or array.
   */
  static constructFromObject(data, obj, itemType) {
    if (Array.isArray(data)) {
      for (var i = 0; i < data.length; i++) {
        if (data.hasOwnProperty(i))
          obj[i] = ApiClient.convertToType(data[i], itemType);
      }
    } else {
      for (var k in data) {
        if (data.hasOwnProperty(k))
          obj[k] = ApiClient.convertToType(data[k], itemType);
      }
    }
  }
}

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The AamParameters model module.
 * @module model/AamParameters
 * @version 1.0.0
 */
class AamParameters {
  /**
   * Constructs a new <code>AamParameters</code>.
   * @alias module:model/AamParameters
   */
  constructor() {
    AamParameters.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>AamParameters</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/AamParameters} obj Optional instance to populate.
   * @return {module:model/AamParameters} The populated <code>AamParameters</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new AamParameters();

      if (data.hasOwnProperty("dcsLocationHint")) {
        obj.dcsLocationHint = ApiClient.convertToType(
          data["dcsLocationHint"],
          "Number"
        );
      }
      if (data.hasOwnProperty("blob")) {
        obj.blob = ApiClient.convertToType(data["blob"], "String");
      }
    }
    return obj;
  }
}

/**
 * @member {Number} dcsLocationHint
 */
AamParameters.prototype.dcsLocationHint = undefined;

/**
 * @member {String} blob
 */
AamParameters.prototype.blob = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */
/**
 * Enum class AuthenticatedState.
 * @enum {}
 * @readonly
 */
class AuthenticatedState {
  constructor() {
    /**
     * value: "unknown"
     * @const
     */
    this["unknown"] = "unknown";

    /**
     * value: "authenticated"
     * @const
     */
    this["authenticated"] = "authenticated";

    /**
     * value: "logged_out"
     * @const
     */
    this["logged_out"] = "logged_out";
  }

  /**
   * Returns a <code>AuthenticatedState</code> enum value from a Javascript object name.
   * @param {Object} data The plain JavaScript object containing the name of the enum value.
   * @return {module:model/AuthenticatedState} The enum <code>AuthenticatedState</code> value.
   */
  static constructFromObject(object) {
    return object;
  }
}

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The MboxOverride model module.
 * @module model/MboxOverride
 * @version 1.0.0
 */
class MboxOverride {
  /**
   * Constructs a new <code>MboxOverride</code>.
   * @alias module:model/MboxOverride
   */
  constructor() {
    MboxOverride.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>MboxOverride</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/MboxOverride} obj Optional instance to populate.
   * @return {module:model/MboxOverride} The populated <code>MboxOverride</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new MboxOverride();

      if (data.hasOwnProperty("ip")) {
        obj.ip = ApiClient.convertToType(data["ip"], "String");
      }
      if (data.hasOwnProperty("time")) {
        obj.time = ApiClient.convertToType(data["time"], "Number");
      }
      if (data.hasOwnProperty("visitorPercentage")) {
        obj.visitorPercentage = ApiClient.convertToType(
          data["visitorPercentage"],
          "Number"
        );
      }
      if (data.hasOwnProperty("aamResponse")) {
        obj.aamResponse = ApiClient.convertToType(
          data["aamResponse"],
          "String"
        );
      }
      if (data.hasOwnProperty("aamSegments")) {
        obj.aamSegments = ApiClient.convertToType(
          data["aamSegments"],
          "String"
        );
      }
    }
    return obj;
  }
}

/**
 * @member {String} ip
 */
MboxOverride.prototype.ip = undefined;

/**
 * @member {Number} time
 */
MboxOverride.prototype.time = undefined;

/**
 * @member {Number} visitorPercentage
 */
MboxOverride.prototype.visitorPercentage = undefined;

/**
 * @member {String} aamResponse
 */
MboxOverride.prototype.aamResponse = undefined;

/**
 * @member {String} aamSegments
 */
MboxOverride.prototype.aamSegments = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The Order model module.
 * @module model/Order
 * @version 1.0.0
 */
class Order {
  /**
   * Constructs a new <code>Order</code>.
   * @alias module:model/Order
   */
  constructor() {
    Order.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>Order</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/Order} obj Optional instance to populate.
   * @return {module:model/Order} The populated <code>Order</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new Order();

      if (data.hasOwnProperty("id")) {
        obj.id = ApiClient.convertToType(data["id"], "String");
      }
      if (data.hasOwnProperty("total")) {
        obj.total = ApiClient.convertToType(data["total"], "Number");
      }
      if (data.hasOwnProperty("purchasedProductIds")) {
        obj.purchasedProductIds = ApiClient.convertToType(
          data["purchasedProductIds"],
          ["String"]
        );
      }
    }
    return obj;
  }
}

/**
 * @member {String} id
 */
Order.prototype.id = undefined;

/**
 * @member {Number} total
 */
Order.prototype.total = undefined;

/**
 * @member {Array.<String>} purchasedProductIds
 */
Order.prototype.purchasedProductIds = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The Parameters model module.
 * @module model/Parameters
 * @version 1.0.0
 */
class Parameters {
  /**
   * Constructs a new <code>Parameters</code>.
   * @alias module:model/Parameters
   * @extends Object
   */
  constructor() {
    Parameters.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>Parameters</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/Parameters} obj Optional instance to populate.
   * @return {module:model/Parameters} The populated <code>Parameters</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new Parameters();

      ApiClient.constructFromObject(data, obj, "String");
    }
    return obj;
  }
}

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The Product model module.
 * @module model/Product
 * @version 1.0.0
 */
class Product {
  /**
   * Constructs a new <code>Product</code>.
   * @alias module:model/Product
   */
  constructor() {
    Product.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>Product</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/Product} obj Optional instance to populate.
   * @return {module:model/Product} The populated <code>Product</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new Product();

      if (data.hasOwnProperty("id")) {
        obj.id = ApiClient.convertToType(data["id"], "String");
      }
      if (data.hasOwnProperty("categoryId")) {
        obj.categoryId = ApiClient.convertToType(data["categoryId"], "String");
      }
    }
    return obj;
  }
}

/**
 * @member {String} id
 */
Product.prototype.id = undefined;

/**
 * @member {String} categoryId
 */
Product.prototype.categoryId = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The RequestLocation model module.
 * @module model/RequestLocation
 * @version 1.0.0
 */
class RequestLocation {
  /**
   * Constructs a new <code>RequestLocation</code>.
   * @alias module:model/RequestLocation
   */
  constructor() {
    RequestLocation.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>RequestLocation</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/RequestLocation} obj Optional instance to populate.
   * @return {module:model/RequestLocation} The populated <code>RequestLocation</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new RequestLocation();

      if (data.hasOwnProperty("pageURL")) {
        obj.pageURL = ApiClient.convertToType(data["pageURL"], "String");
      }
      if (data.hasOwnProperty("referrerURL")) {
        obj.referrerURL = ApiClient.convertToType(
          data["referrerURL"],
          "String"
        );
      }
      if (data.hasOwnProperty("impressionId")) {
        obj.impressionId = ApiClient.convertToType(
          data["impressionId"],
          "String"
        );
      }
    }
    return obj;
  }
}

/**
 * @member {String} pageURL
 */
RequestLocation.prototype.pageURL = undefined;

/**
 * @member {String} referrerURL
 */
RequestLocation.prototype.referrerURL = undefined;

/**
 * @member {String} impressionId
 */
RequestLocation.prototype.impressionId = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The MboxRequest model module.
 * @module model/MboxRequest
 * @version 1.0.0
 */
class MboxRequest {
  /**
   * Constructs a new <code>MboxRequest</code>.
   * @alias module:model/MboxRequest
   * @param indexId {Number}
   * @param mbox {String}
   */
  constructor(indexId, mbox) {
    MboxRequest.initialize(this, indexId, mbox);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj, indexId, mbox) {
    obj["indexId"] = indexId;
    obj["mbox"] = mbox;
  }

  /**
   * Constructs a <code>MboxRequest</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/MboxRequest} obj Optional instance to populate.
   * @return {module:model/MboxRequest} The populated <code>MboxRequest</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new MboxRequest();

      if (data.hasOwnProperty("indexId")) {
        obj.indexId = ApiClient.convertToType(data["indexId"], "Number");
      }
      if (data.hasOwnProperty("mbox")) {
        obj.mbox = ApiClient.convertToType(data["mbox"], "String");
      }
      if (data.hasOwnProperty("parameters")) {
        obj.parameters = Parameters.constructFromObject(data["parameters"]);
      }
      if (data.hasOwnProperty("order")) {
        obj.order = Order.constructFromObject(data["order"]);
      }
      if (data.hasOwnProperty("product")) {
        obj.product = Product.constructFromObject(data["product"]);
      }
      if (data.hasOwnProperty("requestLocation")) {
        obj.requestLocation = RequestLocation.constructFromObject(
          data["requestLocation"]
        );
      }
    }
    return obj;
  }
}

/**
 * @member {Number} indexId
 */
MboxRequest.prototype.indexId = undefined;

/**
 * @member {String} mbox
 */
MboxRequest.prototype.mbox = undefined;

/**
 * @member {module:model/Parameters} parameters
 */
MboxRequest.prototype.parameters = undefined;

/**
 * @member {module:model/Order} order
 */
MboxRequest.prototype.order = undefined;

/**
 * @member {module:model/Product} product
 */
MboxRequest.prototype.product = undefined;

/**
 * @member {module:model/RequestLocation} requestLocation
 */
MboxRequest.prototype.requestLocation = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The QAModePreviewIndex model module.
 * @module model/QAModePreviewIndex
 * @version 1.0.0
 */
class QAModePreviewIndex {
  /**
   * Constructs a new <code>QAModePreviewIndex</code>.
   * @alias module:model/QAModePreviewIndex
   */
  constructor() {
    QAModePreviewIndex.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>QAModePreviewIndex</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/QAModePreviewIndex} obj Optional instance to populate.
   * @return {module:model/QAModePreviewIndex} The populated <code>QAModePreviewIndex</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new QAModePreviewIndex();

      if (data.hasOwnProperty("activityIndex")) {
        obj.activityIndex = ApiClient.convertToType(
          data["activityIndex"],
          "Number"
        );
      }
      if (data.hasOwnProperty("experienceIndex")) {
        obj.experienceIndex = ApiClient.convertToType(
          data["experienceIndex"],
          "Number"
        );
      }
    }
    return obj;
  }
}

/**
 * @member {Number} activityIndex
 */
QAModePreviewIndex.prototype.activityIndex = undefined;

/**
 * @member {Number} experienceIndex
 */
QAModePreviewIndex.prototype.experienceIndex = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The QAMode model module.
 * @module model/QAMode
 * @version 1.0.0
 */
class QAMode {
  /**
   * Constructs a new <code>QAMode</code>.
   * @alias module:model/QAMode
   */
  constructor() {
    QAMode.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>QAMode</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/QAMode} obj Optional instance to populate.
   * @return {module:model/QAMode} The populated <code>QAMode</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new QAMode();

      if (data.hasOwnProperty("token")) {
        obj.token = ApiClient.convertToType(data["token"], "String");
      }
      if (data.hasOwnProperty("listedActivitiesOnly")) {
        obj.listedActivitiesOnly = ApiClient.convertToType(
          data["listedActivitiesOnly"],
          "Boolean"
        );
      }
      if (data.hasOwnProperty("evaluateAsTrueAudienceIds")) {
        obj.evaluateAsTrueAudienceIds = ApiClient.convertToType(
          data["evaluateAsTrueAudienceIds"],
          ["Number"]
        );
      }
      if (data.hasOwnProperty("evaluateAsFalseAudienceIds")) {
        obj.evaluateAsFalseAudienceIds = ApiClient.convertToType(
          data["evaluateAsFalseAudienceIds"],
          ["Number"]
        );
      }
      if (data.hasOwnProperty("previewIndexes")) {
        obj.previewIndexes = ApiClient.convertToType(data["previewIndexes"], [
          QAModePreviewIndex
        ]);
      }
    }
    return obj;
  }
}

/**
 * @member {String} token
 */
QAMode.prototype.token = undefined;

/**
 * @member {Boolean} listedActivitiesOnly
 */
QAMode.prototype.listedActivitiesOnly = undefined;

/**
 * @member {Array.<Number>} evaluateAsTrueAudienceIds
 */
QAMode.prototype.evaluateAsTrueAudienceIds = undefined;

/**
 * @member {Array.<Number>} evaluateAsFalseAudienceIds
 */
QAMode.prototype.evaluateAsFalseAudienceIds = undefined;

/**
 * @member {Array.<module:model/QAModePreviewIndex>} previewIndexes
 */
QAMode.prototype.previewIndexes = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The Trace model module.
 * @module model/Trace
 * @version 1.0.0
 */
class Trace {
  /**
   * Constructs a new <code>Trace</code>.
   * @alias module:model/Trace
   */
  constructor() {
    Trace.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>Trace</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/Trace} obj Optional instance to populate.
   * @return {module:model/Trace} The populated <code>Trace</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new Trace();

      if (data.hasOwnProperty("enabled")) {
        obj.enabled = ApiClient.convertToType(data["enabled"], "Boolean");
      }
    }
    return obj;
  }
}

/**
 * @member {Boolean} enabled
 */
Trace.prototype.enabled = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The CustomerId model module.
 * @module model/CustomerId
 * @version 1.0.0
 */
class CustomerId {
  /**
   * Constructs a new <code>CustomerId</code>.
   * @alias module:model/CustomerId
   */
  constructor() {
    CustomerId.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>CustomerId</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/CustomerId} obj Optional instance to populate.
   * @return {module:model/CustomerId} The populated <code>CustomerId</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new CustomerId();

      if (data.hasOwnProperty("id")) {
        obj.id = ApiClient.convertToType(data["id"], "String");
      }
      if (data.hasOwnProperty("integrationCode")) {
        obj.integrationCode = ApiClient.convertToType(
          data["integrationCode"],
          "String"
        );
      }
      if (data.hasOwnProperty("authenticatedState")) {
        obj.authenticatedState = AuthenticatedState.constructFromObject(
          data["authenticatedState"]
        );
      }
    }
    return obj;
  }
}

/**
 * @member {String} id
 */
CustomerId.prototype.id = undefined;

/**
 * @member {String} integrationCode
 */
CustomerId.prototype.integrationCode = undefined;

/**
 * @member {module:model/AuthenticatedState} authenticatedState
 */
CustomerId.prototype.authenticatedState = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The VisitorId model module.
 * @module model/VisitorId
 * @version 1.0.0
 */
class VisitorId {
  /**
   * Constructs a new <code>VisitorId</code>.
   * @alias module:model/VisitorId
   */
  constructor() {
    VisitorId.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>VisitorId</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/VisitorId} obj Optional instance to populate.
   * @return {module:model/VisitorId} The populated <code>VisitorId</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new VisitorId();

      if (data.hasOwnProperty("tntId")) {
        obj.tntId = ApiClient.convertToType(data["tntId"], "String");
      }
      if (data.hasOwnProperty("thirdPartyId")) {
        obj.thirdPartyId = ApiClient.convertToType(
          data["thirdPartyId"],
          "String"
        );
      }
      if (data.hasOwnProperty("marketingCloudVisitorId")) {
        obj.marketingCloudVisitorId = ApiClient.convertToType(
          data["marketingCloudVisitorId"],
          "String"
        );
      }
      if (data.hasOwnProperty("customerIds")) {
        obj.customerIds = ApiClient.convertToType(data["customerIds"], [
          CustomerId
        ]);
      }
    }
    return obj;
  }
}

/**
 * @member {String} tntId
 */
VisitorId.prototype.tntId = undefined;

/**
 * @member {String} thirdPartyId
 */
VisitorId.prototype.thirdPartyId = undefined;

/**
 * @member {String} marketingCloudVisitorId
 */
VisitorId.prototype.marketingCloudVisitorId = undefined;

/**
 * @member {Array.<module:model/CustomerId>} customerIds
 */
VisitorId.prototype.customerIds = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The BatchMboxRequest model module.
 * @module model/BatchMboxRequest
 * @version 1.0.0
 */
class BatchMboxRequest {
  /**
   * Constructs a new <code>BatchMboxRequest</code>.
   * @alias module:model/BatchMboxRequest
   * @param client {String}
   */
  constructor(client) {
    BatchMboxRequest.initialize(this, client);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj, client) {
    obj["client"] = client;
  }

  /**
   * Constructs a <code>BatchMboxRequest</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/BatchMboxRequest} obj Optional instance to populate.
   * @return {module:model/BatchMboxRequest} The populated <code>BatchMboxRequest</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new BatchMboxRequest();

      if (data.hasOwnProperty("client")) {
        obj.client = ApiClient.convertToType(data["client"], "String");
      }
      if (data.hasOwnProperty("id")) {
        obj.id = VisitorId.constructFromObject(data["id"]);
      }
      if (data.hasOwnProperty("aamParameters")) {
        obj.aamParameters = AamParameters.constructFromObject(
          data["aamParameters"]
        );
      }
      if (data.hasOwnProperty("environmentId")) {
        obj.environmentId = ApiClient.convertToType(
          data["environmentId"],
          "Number"
        );
      }
      if (data.hasOwnProperty("host")) {
        obj.host = ApiClient.convertToType(data["host"], "String");
      }
      if (data.hasOwnProperty("ipAddress")) {
        obj.ipAddress = ApiClient.convertToType(data["ipAddress"], "String");
      }
      if (data.hasOwnProperty("contentAsJson")) {
        obj.contentAsJson = ApiClient.convertToType(
          data["contentAsJson"],
          "Boolean"
        );
      }
      if (data.hasOwnProperty("supplementalDataId")) {
        obj.supplementalDataId = ApiClient.convertToType(
          data["supplementalDataId"],
          "String"
        );
      }
      if (data.hasOwnProperty("profileParameters")) {
        obj.profileParameters = Parameters.constructFromObject(
          data["profileParameters"]
        );
      }
      if (data.hasOwnProperty("mboxes")) {
        obj.mboxes = ApiClient.convertToType(data["mboxes"], [MboxRequest]);
      }
      if (data.hasOwnProperty("mboxOverride")) {
        obj.mboxOverride = MboxOverride.constructFromObject(
          data["mboxOverride"]
        );
      }
      if (data.hasOwnProperty("qaMode")) {
        obj.qaMode = QAMode.constructFromObject(data["qaMode"]);
      }
      if (data.hasOwnProperty("trace")) {
        obj.trace = Trace.constructFromObject(data["trace"]);
      }
    }
    return obj;
  }
}

/**
 * @member {String} client
 */
BatchMboxRequest.prototype.client = undefined;

/**
 * @member {module:model/VisitorId} id
 */
BatchMboxRequest.prototype.id = undefined;

/**
 * @member {module:model/AamParameters} aamParameters
 */
BatchMboxRequest.prototype.aamParameters = undefined;

/**
 * @member {Number} environmentId
 */
BatchMboxRequest.prototype.environmentId = undefined;

/**
 * @member {String} host
 */
BatchMboxRequest.prototype.host = undefined;

/**
 * @member {String} ipAddress
 */
BatchMboxRequest.prototype.ipAddress = undefined;

/**
 * @member {Boolean} contentAsJson
 */
BatchMboxRequest.prototype.contentAsJson = undefined;

/**
 * @member {String} supplementalDataId
 */
BatchMboxRequest.prototype.supplementalDataId = undefined;

/**
 * @member {module:model/Parameters} profileParameters
 */
BatchMboxRequest.prototype.profileParameters = undefined;

/**
 * @member {Array.<module:model/MboxRequest>} mboxes
 */
BatchMboxRequest.prototype.mboxes = undefined;

/**
 * @member {module:model/MboxOverride} mboxOverride
 */
BatchMboxRequest.prototype.mboxOverride = undefined;

/**
 * @member {module:model/QAMode} qaMode
 */
BatchMboxRequest.prototype.qaMode = undefined;

/**
 * @member {module:model/Trace} trace
 */
BatchMboxRequest.prototype.trace = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */
/**
 * Enum class ErrorType.
 * @enum {}
 * @readonly
 */
class ErrorType {
  constructor() {
    /**
     * value: "unknown"
     * @const
     */
    this["unknown"] = "unknown";

    /**
     * value: "time_out"
     * @const
     */
    this["time_out"] = "time_out";

    /**
     * value: "invalid_event_token"
     * @const
     */
    this["invalid_event_token"] = "invalid_event_token";

    /**
     * value: "invalid_profile_script_token"
     * @const
     */
    this["invalid_profile_script_token"] = "invalid_profile_script_token";

    /**
     * value: "mbox_blocked"
     * @const
     */
    this["mbox_blocked"] = "mbox_blocked";

    /**
     * value: "invalid_campaign_request"
     * @const
     */
    this["invalid_campaign_request"] = "invalid_campaign_request";

    /**
     * value: "invalid_json"
     * @const
     */
    this["invalid_json"] = "invalid_json";

    /**
     * value: "content_generation"
     * @const
     */
    this["content_generation"] = "content_generation";

    /**
     * value: "unsupported_offer_type"
     * @const
     */
    this["unsupported_offer_type"] = "unsupported_offer_type";
  }

  /**
   * Returns a <code>ErrorType</code> enum value from a Javascript object name.
   * @param {Object} data The plain JavaScript object containing the name of the enum value.
   * @return {module:model/ErrorType} The enum <code>ErrorType</code> value.
   */
  static constructFromObject(object) {
    return object;
  }
}

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The ResponseTokens model module.
 * @module model/ResponseTokens
 * @version 1.0.0
 */
class ResponseTokens {
  /**
   * Constructs a new <code>ResponseTokens</code>.
   * @alias module:model/ResponseTokens
   * @extends Object
   */
  constructor() {
    ResponseTokens.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>ResponseTokens</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/ResponseTokens} obj Optional instance to populate.
   * @return {module:model/ResponseTokens} The populated <code>ResponseTokens</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new ResponseTokens();

      ApiClient.constructFromObject(data, obj, "Object");
    }
    return obj;
  }
}

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The MboxResponse model module.
 * @module model/MboxResponse
 * @version 1.0.0
 */
class MboxResponse {
  /**
   * Constructs a new <code>MboxResponse</code>.
   * @alias module:model/MboxResponse
   * @param mbox {String}
   */
  constructor(mbox) {
    MboxResponse.initialize(this, mbox);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj, mbox) {
    obj["mbox"] = mbox;
  }

  /**
   * Constructs a <code>MboxResponse</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/MboxResponse} obj Optional instance to populate.
   * @return {module:model/MboxResponse} The populated <code>MboxResponse</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new MboxResponse();

      if (data.hasOwnProperty("mbox")) {
        obj.mbox = ApiClient.convertToType(data["mbox"], "String");
      }
      if (data.hasOwnProperty("parameters")) {
        obj.parameters = Parameters.constructFromObject(data["parameters"]);
      }
      if (data.hasOwnProperty("order")) {
        obj.order = Order.constructFromObject(data["order"]);
      }
      if (data.hasOwnProperty("product")) {
        obj.product = Product.constructFromObject(data["product"]);
      }
      if (data.hasOwnProperty("content")) {
        obj.content = ApiClient.convertToType(data["content"], Object);
      }
      if (data.hasOwnProperty("responseTokens")) {
        obj.responseTokens = ResponseTokens.constructFromObject(
          data["responseTokens"]
        );
      }
      if (data.hasOwnProperty("clickToken")) {
        obj.clickToken = ApiClient.convertToType(data["clickToken"], "String");
      }
      if (data.hasOwnProperty("clientSideAnalyticsLoggingPayload")) {
        obj.clientSideAnalyticsLoggingPayload = ApiClient.convertToType(
          data["clientSideAnalyticsLoggingPayload"],
          "String"
        );
      }
      if (data.hasOwnProperty("errorType")) {
        obj.errorType = ErrorType.constructFromObject(data["errorType"]);
      }
      if (data.hasOwnProperty("trace")) {
        obj.trace = ApiClient.convertToType(data["trace"], Object);
      }
    }
    return obj;
  }
}

/**
 * @member {String} mbox
 */
MboxResponse.prototype.mbox = undefined;

/**
 * @member {module:model/Parameters} parameters
 */
MboxResponse.prototype.parameters = undefined;

/**
 * @member {module:model/Order} order
 */
MboxResponse.prototype.order = undefined;

/**
 * @member {module:model/Product} product
 */
MboxResponse.prototype.product = undefined;

/**
 * @member {Object} content
 */
MboxResponse.prototype.content = undefined;

/**
 * @member {module:model/ResponseTokens} responseTokens
 */
MboxResponse.prototype.responseTokens = undefined;

/**
 * @member {String} clickToken
 */
MboxResponse.prototype.clickToken = undefined;

/**
 * @member {String} clientSideAnalyticsLoggingPayload
 */
MboxResponse.prototype.clientSideAnalyticsLoggingPayload = undefined;

/**
 * @member {module:model/ErrorType} errorType
 */
MboxResponse.prototype.errorType = undefined;

/**
 * @member {Object} trace
 */
MboxResponse.prototype.trace = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The BatchMboxResponse model module.
 * @module model/BatchMboxResponse
 * @version 1.0.0
 */
class BatchMboxResponse {
  /**
   * Constructs a new <code>BatchMboxResponse</code>.
   * @alias module:model/BatchMboxResponse
   */
  constructor() {
    BatchMboxResponse.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj) {}

  /**
   * Constructs a <code>BatchMboxResponse</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/BatchMboxResponse} obj Optional instance to populate.
   * @return {module:model/BatchMboxResponse} The populated <code>BatchMboxResponse</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new BatchMboxResponse();

      if (data.hasOwnProperty("requestId")) {
        obj.requestId = ApiClient.convertToType(data["requestId"], "String");
      }
      if (data.hasOwnProperty("client")) {
        obj.client = ApiClient.convertToType(data["client"], "String");
      }
      if (data.hasOwnProperty("id")) {
        obj.id = VisitorId.constructFromObject(data["id"]);
      }
      if (data.hasOwnProperty("edgeHost")) {
        obj.edgeHost = ApiClient.convertToType(data["edgeHost"], "String");
      }
      if (data.hasOwnProperty("contentAsJson")) {
        obj.contentAsJson = ApiClient.convertToType(
          data["contentAsJson"],
          "Boolean"
        );
      }
      if (data.hasOwnProperty("mboxResponses")) {
        obj.mboxResponses = ApiClient.convertToType(data["mboxResponses"], [
          MboxResponse
        ]);
      }
    }
    return obj;
  }
}

/**
 * @member {String} requestId
 */
BatchMboxResponse.prototype.requestId = undefined;

/**
 * @member {String} client
 */
BatchMboxResponse.prototype.client = undefined;

/**
 * @member {module:model/VisitorId} id
 */
BatchMboxResponse.prototype.id = undefined;

/**
 * @member {String} edgeHost
 */
BatchMboxResponse.prototype.edgeHost = undefined;

/**
 * @member {Boolean} contentAsJson
 */
BatchMboxResponse.prototype.contentAsJson = undefined;

/**
 * @member {Array.<module:model/MboxResponse>} mboxResponses
 */
BatchMboxResponse.prototype.mboxResponses = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * The UnexpectedError model module.
 * @module model/UnexpectedError
 * @version 1.0.0
 */
class UnexpectedError {
  /**
   * Constructs a new <code>UnexpectedError</code>.
   * @alias module:model/UnexpectedError
   * @param status {Number}
   * @param message {String}
   */
  constructor(status, message) {
    UnexpectedError.initialize(this, status, message);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  static initialize(obj, status, message) {
    obj["status"] = status;
    obj["message"] = message;
  }

  /**
   * Constructs a <code>UnexpectedError</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/UnexpectedError} obj Optional instance to populate.
   * @return {module:model/UnexpectedError} The populated <code>UnexpectedError</code> instance.
   */
  static constructFromObject(data, obj) {
    if (data) {
      obj = obj || new UnexpectedError();

      if (data.hasOwnProperty("status")) {
        obj.status = ApiClient.convertToType(data["status"], "Number");
      }
      if (data.hasOwnProperty("message")) {
        obj.message = ApiClient.convertToType(data["message"], "String");
      }
    }
    return obj;
  }
}

/**
 * @member {Number} status
 */
UnexpectedError.prototype.status = undefined;

/**
 * @member {String} message
 */
UnexpectedError.prototype.message = undefined;

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

/**
 * TargetBatchMbox service.
 * @module api/TargetBatchMboxApi
 * @version 1.0.0
 */
class TargetBatchMboxApi {
  /**
   * Constructs a new TargetBatchMboxApi.
   * @alias module:api/TargetBatchMboxApi
   * @class
   * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
   * default to {@link module:ApiClient#instance} if unspecified.
   */
  constructor(apiClient) {
    this.apiClient = apiClient || ApiClient.instance;
  }

  /**
   * Retrieve Target offers
   * @param {module:model/BatchMboxRequest} batchMboxRequest Batch mbox request
   * @param {Object} opts Optional parameters
   * @param {String} opts.client The customer client code
   * @param {String} opts.sessionId The user session ID
   * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data of type {@link module:model/BatchMboxResponse} and HTTP response
   */
  executeWithHttpInfo(batchMboxRequest, opts) {
    opts = opts || {};
    const postBody = batchMboxRequest;

    // verify the required parameter 'batchMboxRequest' is set
    if (batchMboxRequest === undefined || batchMboxRequest === null) {
      throw new Error(
        "Missing the required parameter 'batchMboxRequest' when calling execute"
      );
    }

    const host = opts["host"];
    const queryParams = {
      client: opts["client"],
      sessionId: opts["sessionId"]
    };
    const headerParams = Object.assign({}, opts["headers"]);
    const returnType = BatchMboxResponse;

    return this.apiClient.callApi(
      host,
      "/rest/v2/batchmbox",
      queryParams,
      headerParams,
      postBody,
      returnType
    );
  }

  /**
   * Retrieve Target offers
   * @param {module:model/BatchMboxRequest} batchMboxRequest Batch mbox request
   * @param {Object} opts Optional parameters
   * @param {String} opts.client The customer client code
   * @param {String} opts.sessionId The user session ID
   * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/BatchMboxResponse}
   */
  execute(batchMboxRequest, opts) {
    return this.executeWithHttpInfo(batchMboxRequest, opts);
  }
}

/**
 * Target batch mbox client
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * OpenAPI spec version: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */

exports.ApiClient = ApiClient;
exports.AamParameters = AamParameters;
exports.AuthenticatedState = AuthenticatedState;
exports.BatchMboxRequest = BatchMboxRequest;
exports.BatchMboxResponse = BatchMboxResponse;
exports.CustomerId = CustomerId;
exports.ErrorType = ErrorType;
exports.MboxOverride = MboxOverride;
exports.MboxRequest = MboxRequest;
exports.MboxResponse = MboxResponse;
exports.Order = Order;
exports.Parameters = Parameters;
exports.Product = Product;
exports.QAMode = QAMode;
exports.QAModePreviewIndex = QAModePreviewIndex;
exports.RequestLocation = RequestLocation;
exports.ResponseTokens = ResponseTokens;
exports.Trace = Trace;
exports.UnexpectedError = UnexpectedError;
exports.VisitorId = VisitorId;
exports.TargetBatchMboxApi = TargetBatchMboxApi;
