// @adobe/target-node-client/lib/common/utils.js
/** ***********************************************************************
 * ADOBE CONFIDENTIAL
 * ___________________
 *
 *  Copyright 2017 Adobe Systems Incorporated
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Adobe Systems Incorporated and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Adobe Systems Incorporated and its
 * suppliers and are protected by all applicable intellectual property
 * laws, including trade secret and copyright laws.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe Systems Incorporated.
 ************************************************************************* */
const _ = require("lodash");
const Visitor = require("@adobe-mcid/visitor-js-server");

const NOOP_LOGGER = {
  log() {}
};

function getLogger(options) {
  if (!options.logger) {
    return NOOP_LOGGER;
  }

  const { logger } = options;

  if (!_.isFunction(logger.log)) {
    return NOOP_LOGGER;
  }

  return logger;
}

function isEmptyObject(value) {
  if (_.isEmpty(value)) {
    return true;
  }

  return _.values(value).filter(v => !_.isUndefined(v)).length === 0;
}

function copyObjectByFields(fields, sourceObj) {
  if (!_.isArray(fields)) {
    return {};
  }

  return fields.reduce((accumulator, field) => {
    const result = {};
    if (!_.isUndefined(sourceObj[field])) {
      result[field] = sourceObj[field];
    }
    return Object.assign({}, accumulator, result);
  }, {});
}

function throwError(msg) {
  throw new Error(msg);
}

function normalizeCustomerIds(customerIds) {
  const result = _.keys(customerIds).reduce((acc, key) => {
    const customerId = customerIds[key];

    if (!customerId) {
      return acc;
    }

    acc[key] = acc[key] || {};

    // eslint-disable-next-line eqeqeq
    if (customerId.id != undefined) {
      acc[key].id = customerId.id;
    } else {
      acc[key].id = customerId;
    }

    // eslint-disable-next-line eqeqeq
    if (customerId.authState != undefined) {
      acc[key].authState = customerId.authState;
    } else {
      acc[key].authState = Visitor.AuthState.UNKNOWN;
    }

    return acc;
  }, {});

  return isEmptyObject(result) ? null : result;
}

function createTargetOptions(visitor, config, logger, request) {
  const { targetCookie, targetLocationHintCookie, customerIds } = request;
  const result = {
    visitor,
    config,
    logger,
    targetCookie,
    targetLocationHintCookie
  };

  if (customerIds) {
    result.customerIds = normalizeCustomerIds(customerIds);
  }

  return result;
}

function createVisitor(request, config) {
  const { organizationId } = config;
  const { visitorCookie, customerIds } = request;
  const result = request.visitor || new Visitor(organizationId, visitorCookie);

  if (customerIds) {
    result.setCustomerIDs(customerIds);
  }

  return result;
}

function createResponse(visitor, response) {
  const visitorState = { visitorState: visitor.getState() };

  return Object.assign({}, visitorState, response);
}

module.exports = {
  isEmptyObject,
  throwError,
  copyObjectByFields,
  getLogger,
  createVisitor,
  createTargetOptions,
  createResponse
};
