// @adobe/target-node-client/lib/messages.js
/** ***********************************************************************
 * ADOBE CONFIDENTIAL
 * ___________________
 *
 *  Copyright 2017 Adobe Systems Incorporated
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Adobe Systems Incorporated and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Adobe Systems Incorporated and its
 * suppliers and are protected by all applicable intellectual property
 * laws, including trade secret and copyright laws.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe Systems Incorporated.
 ************************************************************************* */

module.exports = {
  ORG_ID_REQUIRED: "Organization ID is required",
  CONSUMER_ID_REQUIRED: "Consumer ID is required",
  REQUEST_REQUIRED: "Request is required",
  MBOX_REQUIRED: "Mbox is required",
  CLIENT_REQUIRED: "Client is required",
  VISITOR_REQUIRED: "Visitor is required",
  CONFIG_REQUIRED: "Config is required",
  LOGGER_REQUIRED: "Logger is required",
  MBOXES_REQUIRED: "The mboxes array is required"
};
